onnx/gen_proto.py

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env python
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import os
+import re
+
+autogen_header = """\
+//
+// WARNING: This file is automatically generated!  Please edit onnx.proto.in.
+//
+
+
+"""
+
+IF_ONNX_ML_REGEX = re.compile(r'\s*//\s*#if\s+ONNX-ML\s*$')
+ENDIF_ONNX_ML_REGEX = re.compile(r'\s*//\s*#endif\s*$')
+
+def process_ifs(lines, onnx_ml):
+    in_if = False
+    for line in lines:
+        if IF_ONNX_ML_REGEX.match(line):
+            assert not in_if
+            in_if = True
+        elif ENDIF_ONNX_ML_REGEX.match(line):
+            assert in_if
+            in_if = False
+        else:
+            if not in_if or (in_if and onnx_ml):
+                yield line
+
+PROTO_SYNTAX_REGEX = re.compile(r'(\s*)syntax\s*=\s*"proto2"\s*;\s*$')
+OPTIONAL_REGEX = re.compile(r'(\s*)optional(\s.*)$')
+
+def convert_to_proto3(lines):
+    for line in lines:
+        # Set the syntax specifier
+        m = PROTO_SYNTAX_REGEX.match(line)
+        if m:
+            yield m.group(1) + 'syntax = "proto3";'
+            continue
+
+        # Remove optinoal keywords
+        m = OPTIONAL_REGEX.match(line)
+        if m:
+            yield m.group(1) + m.group(2)
```

Comment:
nit: when "optional" is stripped the trailing whitespace is left behind causing different indentation for proto3 vs proto2 versions.